### PR TITLE
Fix Oauth2ApiClient token refresh, responses

### PR DESCRIPTION
**Pull request by nonword**
Fixes various things with Oauth2ApiClient:
 - Ensure token is refetched when it expires
 - HTTP functions should return the `requests` response object, not just the json
 - Raise errors based on status_code
 - Better test coverage

### CODE REVIEW

File: src/nypl_py_utils/classes/oauth2_api_client.py [hunk 1]
```diff
@@ -56,18 +55,30 @@ def _do_http_method(self, method, request_path, **kwargs):
...
+            if '_do_http_method_token_refreshes' not in kwargs.keys():
+                kwargs['_do_http_method_token_refreshes'] = 1
+            else:
+                kwargs['_do_http_method_token_refreshes'] += 1
+                if kwargs['_do_http_method_token_refreshes'] > 3:
+                    raise Exception('Exhausted token refreshes')
 
+            self._generate_access_token()
             return self._do_http_method(method, request_path, **kwargs)
         except TimeoutError as e:
```
Comment (aaronfriedman6) [inline src/nypl_py_utils/classes/oauth2_api_client.py:83]: So this exception is non-blocking?

Comment (nonword) [reply src/nypl_py_utils/classes/oauth2_api_client.py]: We shouldn't be catching `TimeoutError`s and returning an empty dict as done here. Like other errors, we should allow the integrating app to catch it. Removed the catch. Is that what you mean?

File: src/nypl_py_utils/classes/oauth2_api_client.py [hunk 2]
```diff
@@ -56,18 +55,30 @@ def _do_http_method(self, method, request_path, **kwargs):
...
+            return resp
+        except TokenExpiredError:
+            self.logger.debug('TokenExpiredError encountered')
+
+            if '_do_http_method_token_refreshes' not in kwargs.keys():
+                kwargs['_do_http_method_token_refreshes'] = 1
+            else:
+                kwargs['_do_http_method_token_refreshes'] += 1
+                if kwargs['_do_http_method_token_refreshes'] > 3:
+                    raise Exception('Exhausted token refreshes')
```
[the remark this answers is not in the document]
Comment (nonword) [reply src/nypl_py_utils/classes/oauth2_api_client.py]: Done. Thanks

File: src/nypl_py_utils/classes/oauth2_api_client.py [hunk 3]
```diff
@@ -56,18 +55,30 @@ def _do_http_method(self, method, request_path, **kwargs):
...
+            # Build kwargs cleaned of local variables:
+            kwargs_cleaned = {k: kwargs[k] for k in kwargs
+                              if not k.startswith('_do_http_method_')}
+            resp = self.oauth_client.request(method, url, **kwargs_cleaned)
+            resp.raise_for_status()
+            return resp
+        except TokenExpiredError:
+            self.logger.debug('TokenExpiredError encountered')
+
+            if '_do_http_method_token_refreshes' not in kwargs.keys():
```
[the remark this answers is not in the document]
Comment (nonword) [reply src/nypl_py_utils/classes/oauth2_api_client.py]: Much nicer. Thanks

File: src/nypl_py_utils/classes/oauth2_api_client.py (hunk 1 above)
Comment (aaronfriedman6) [reply src/nypl_py_utils/classes/oauth2_api_client.py]: You might catch it and raise the custom Oauth2ApiClientError or leave it as throwing a TimeoutError -- up to you

**Pull request merged**